// de-loggingsystem/lib/errorcode.js
/**
 * Error codes
 */
export var ErrorCode;
(function (ErrorCode) {
    ErrorCode[ErrorCode["OK"] = 0] = "OK";
    ErrorCode[ErrorCode["Error"] = 1] = "Error";
    ErrorCode[ErrorCode["DocumentUndefined"] = 2] = "DocumentUndefined";
    ErrorCode[ErrorCode["NoName"] = 3] = "NoName";
    ErrorCode[ErrorCode["EngineInitialization"] = 100] = "EngineInitialization";
    ErrorCode[ErrorCode["EngineFailed"] = 101] = "EngineFailed";
    ErrorCode[ErrorCode["EngineInstanceNull"] = 102] = "EngineInstanceNull";
    ErrorCode[ErrorCode["EngineInstanceNotNull"] = 103] = "EngineInstanceNotNull";
    ErrorCode[ErrorCode["EngineInstanceUndefined"] = 104] = "EngineInstanceUndefined";
    ErrorCode[ErrorCode["EngineClientNotSet"] = 105] = "EngineClientNotSet";
    ErrorCode[ErrorCode["EngineWindowUndefined"] = 106] = "EngineWindowUndefined";
    ErrorCode[ErrorCode["EngineStartedEarly"] = 107] = "EngineStartedEarly";
    ErrorCode[ErrorCode["EngineRunning"] = 108] = "EngineRunning";
    ErrorCode[ErrorCode["EngineNotRunning"] = 109] = "EngineNotRunning";
    ErrorCode[ErrorCode["EngineCleanupFailed"] = 110] = "EngineCleanupFailed";
    ErrorCode[ErrorCode["ContainerUndefined"] = 111] = "ContainerUndefined";
    ErrorCode[ErrorCode["WebGLNotInitialized"] = 112] = "WebGLNotInitialized";
    ErrorCode[ErrorCode["DataTypeNotRecognized"] = 113] = "DataTypeNotRecognized";
    ErrorCode[ErrorCode["MessageSystemInitialization"] = 200] = "MessageSystemInitialization";
    ErrorCode[ErrorCode["MessageSystemUndefined"] = 201] = "MessageSystemUndefined";
    ErrorCode[ErrorCode["DuplicateListener"] = 202] = "DuplicateListener";
    ErrorCode[ErrorCode["UnsubscribeFailed"] = 203] = "UnsubscribeFailed";
    ErrorCode[ErrorCode["MessageRecieverNotFound"] = 204] = "MessageRecieverNotFound";
    ErrorCode[ErrorCode["FailedAddingListener"] = 205] = "FailedAddingListener";
    ErrorCode[ErrorCode["ListenerUndefined"] = 206] = "ListenerUndefined";
    ErrorCode[ErrorCode["BrowserWindowUndefined"] = 300] = "BrowserWindowUndefined";
    ErrorCode[ErrorCode["BrowserWindowDidNotClose"] = 301] = "BrowserWindowDidNotClose";
    ErrorCode[ErrorCode["ZoneUndefined"] = 400] = "ZoneUndefined";
    ErrorCode[ErrorCode["ZoneHasNoObjects"] = 401] = "ZoneHasNoObjects";
    ErrorCode[ErrorCode["ZoneDoesNotExist"] = 402] = "ZoneDoesNotExist";
    ErrorCode[ErrorCode["ZoneID"] = 403] = "ZoneID";
    ErrorCode[ErrorCode["ZoneDescription"] = 404] = "ZoneDescription";
    ErrorCode[ErrorCode["SceneUndefined"] = 405] = "SceneUndefined";
    ErrorCode[ErrorCode["SceneNameUndefined"] = 406] = "SceneNameUndefined";
    ErrorCode[ErrorCode["SceneManagerUndefined"] = 407] = "SceneManagerUndefined";
    ErrorCode[ErrorCode["SceneManagerCleanupFailed"] = 408] = "SceneManagerCleanupFailed";
    ErrorCode[ErrorCode["EntityInitialization"] = 500] = "EntityInitialization";
    ErrorCode[ErrorCode["EntityParentUndefined"] = 501] = "EntityParentUndefined";
    ErrorCode[ErrorCode["EntityAlreadyHasChild"] = 502] = "EntityAlreadyHasChild";
    ErrorCode[ErrorCode["EntityAlreadyHasComponent"] = 503] = "EntityAlreadyHasComponent";
    ErrorCode[ErrorCode["EntityChildNotFound"] = 504] = "EntityChildNotFound";
    ErrorCode[ErrorCode["EntityComponentNotFound"] = 505] = "EntityComponentNotFound";
    ErrorCode[ErrorCode["NoPosition"] = 506] = "NoPosition";
    ErrorCode[ErrorCode["NoStartPosition"] = 507] = "NoStartPosition";
    ErrorCode[ErrorCode["NoEndPosition"] = 508] = "NoEndPosition";
    ErrorCode[ErrorCode["NoDirection"] = 509] = "NoDirection";
    ErrorCode[ErrorCode["NoPlayerCollisionComponentName"] = 510] = "NoPlayerCollisionComponentName";
    ErrorCode[ErrorCode["NoGroundCollisionComponentName"] = 511] = "NoGroundCollisionComponentName";
    ErrorCode[ErrorCode["NoEnemyCollisionComponentName"] = 512] = "NoEnemyCollisionComponentName";
    ErrorCode[ErrorCode["NoFlagCollisionComponentName"] = 513] = "NoFlagCollisionComponentName";
    ErrorCode[ErrorCode["RenderSystemUndefined"] = 600] = "RenderSystemUndefined";
    ErrorCode[ErrorCode["RenderSystemInitializationFailed"] = 601] = "RenderSystemInitializationFailed";
    ErrorCode[ErrorCode["CanvasNotFound"] = 602] = "CanvasNotFound";
    ErrorCode[ErrorCode["CanvasNotDefined"] = 603] = "CanvasNotDefined";
    ErrorCode[ErrorCode["ShaderType"] = 604] = "ShaderType";
    ErrorCode[ErrorCode["ShaderLoad"] = 605] = "ShaderLoad";
    ErrorCode[ErrorCode["WebGLBuffer"] = 606] = "WebGLBuffer";
    ErrorCode[ErrorCode["WebGLBufferNotCreated"] = 607] = "WebGLBufferNotCreated";
    ErrorCode[ErrorCode["WebGLProgram"] = 608] = "WebGLProgram";
    ErrorCode[ErrorCode["RenderSystemCleanupFailed"] = 609] = "RenderSystemCleanupFailed";
    ErrorCode[ErrorCode["SceneNotFound"] = 610] = "SceneNotFound";
    ErrorCode[ErrorCode["SceneNotActive"] = 611] = "SceneNotActive";
    ErrorCode[ErrorCode["SceneTimedOut"] = 612] = "SceneTimedOut";
    ErrorCode[ErrorCode["SpriteNotAttached"] = 613] = "SpriteNotAttached";
    ErrorCode[ErrorCode["GLUtilityNotInitialized"] = 614] = "GLUtilityNotInitialized";
    ErrorCode[ErrorCode["NoAnimatedSpriteName"] = 615] = "NoAnimatedSpriteName";
    ErrorCode[ErrorCode["NoFrameWidth"] = 616] = "NoFrameWidth";
    ErrorCode[ErrorCode["NoFrameHeight"] = 617] = "NoFrameHeight";
    ErrorCode[ErrorCode["NoFrameCount"] = 618] = "NoFrameCount";
    ErrorCode[ErrorCode["NoFrameSequence"] = 619] = "NoFrameSequence";
    ErrorCode[ErrorCode["PhysicsSystemUndefined"] = 700] = "PhysicsSystemUndefined";
    ErrorCode[ErrorCode["NoShape"] = 701] = "NoShape";
    ErrorCode[ErrorCode["NoShapeType"] = 702] = "NoShapeType";
    ErrorCode[ErrorCode["ShapeNotAllowed"] = 703] = "ShapeNotAllowed";
    ErrorCode[ErrorCode["NoRadius"] = 704] = "NoRadius";
    ErrorCode[ErrorCode["NoWidth"] = 705] = "NoWidth";
    ErrorCode[ErrorCode["NoHeight"] = 706] = "NoHeight";
    ErrorCode[ErrorCode["SoundSystemUndefined"] = 900] = "SoundSystemUndefined";
    ErrorCode[ErrorCode["SoundName"] = 901] = "SoundName";
    ErrorCode[ErrorCode["WindowUndefined"] = 900] = "WindowUndefined";
    ErrorCode[ErrorCode["GameWindowUndefined"] = 901] = "GameWindowUndefined";
    ErrorCode[ErrorCode["NotElectronWindow"] = 902] = "NotElectronWindow";
    ErrorCode[ErrorCode["IOSystemUndefined"] = 1000] = "IOSystemUndefined";
    ErrorCode[ErrorCode["AssetLoaderUninitialized"] = 1100] = "AssetLoaderUninitialized";
    ErrorCode[ErrorCode["AssetManagerUndefined"] = 1101] = "AssetManagerUndefined";
    ErrorCode[ErrorCode["NoFileExtension"] = 1102] = "NoFileExtension";
    ErrorCode[ErrorCode["JSONDataUndefined"] = 1103] = "JSONDataUndefined";
    ErrorCode[ErrorCode["LoadAssetFailed"] = 1104] = "LoadAssetFailed";
    ErrorCode[ErrorCode["NoAssetName"] = 1105] = "NoAssetName";
    ErrorCode[ErrorCode["FileTypeNotAcceptable"] = 1106] = "FileTypeNotAcceptable";
    ErrorCode[ErrorCode["AssetManagerDidNotGetAsset"] = 1107] = "AssetManagerDidNotGetAsset";
    ErrorCode[ErrorCode["TextureNotLoaded"] = 1108] = "TextureNotLoaded";
    ErrorCode[ErrorCode["ErrorLoadingFile"] = 1109] = "ErrorLoadingFile";
    ErrorCode[ErrorCode["ReadJSONFile"] = 1110] = "ReadJSONFile";
    ErrorCode[ErrorCode["WriteJSONFile"] = 1111] = "WriteJSONFile";
    ErrorCode[ErrorCode["FileContentsNotRead"] = 1112] = "FileContentsNotRead";
    ErrorCode[ErrorCode["MissingTypeBuilder"] = 1113] = "MissingTypeBuilder";
    ErrorCode[ErrorCode["TypeUndefined"] = 1114] = "TypeUndefined";
    ErrorCode[ErrorCode["NoActions"] = 1115] = "NoActions";
    ErrorCode[ErrorCode["NoCursor"] = 1116] = "NoCursor";
    ErrorCode[ErrorCode["NoButtons"] = 1117] = "NoButtons";
})(ErrorCode || (ErrorCode = {}));
