// de-loggingsystem/lib/loglevel.js
/**
 * Logging levels
 *
 * * Info is for developer and player information; general warnings and calls.
 * * Debug is for developer information only; for debugging purposes and not shipped in the final version.
 * * Warning is for developer and player; this is information that is handled, but should not be occuring normally.
 * * Error is for developer and player; this is for engine errors that are handled but will cause weird
 * results with the engine's running.
 * * Critical is for developer and player; this is when the engine will crash. Errors here should throw an error, or
 * handle an error and throw a new one. This is when the engine cannot continue running in any curcumstance.
 */
export var LogLevel;
(function (LogLevel) {
    LogLevel[LogLevel["info"] = 0] = "info";
    LogLevel[LogLevel["debug"] = 1] = "debug";
    LogLevel[LogLevel["warning"] = 2] = "warning";
    LogLevel[LogLevel["error"] = 3] = "error";
    LogLevel[LogLevel["critical"] = 4] = "critical";
})(LogLevel || (LogLevel = {}));
